### PR TITLE
✨ Allowing update of OpenstackCluster API server fixed IP

### LINKED ISSUE #1589: Support update of OpenstackCluster's apiServerFixedIP
/kind feature

**Describe the solution you'd like**

At the current state (5ac49e7475de3c9b95df3d73208b79319fc322bc) there's no chance to update the `openstackcluster.spec.apiServerFixedIP` value once a resource has been created.

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/5ac49e7475de3c9b95df3d73208b79319fc322bc/api/v1alpha7/openstackcluster_webhook.go#L68-L127

It would be great if the value could be updated in certain circumstances, such as:
- a floating IP is disabled
- the value is still empty

With such an approach, we're going to preserve the same expected behaviour without breaking any previous expectations.

**Anything else you would like to add:**

I'm working on integrating [Kamaji](http://github.com/clastix/kamaji) with Cluster API and the community demanded to start integrating it with OpenStack.

Right now we found a consensus in writing the first implementation of a [Control Plane Provider](https://github.com/clastix/cluster-api-control-plane-provider-kamaji/discussions/1) by finding a consensus and a workflow, besides some bending in the workflow and state-machine contracts.

tl;dr; by workaround this blocker by allowing the update of the `spec.apiServerFixedIP`, I was able to create OpenStack machines connected to a remote Control Plane hosted in another cloud provider.

```
NAME              PHASE         AGE     VERSION
capi-quickstart   Provisioned   5h37m   
```

```
NAME                                                        READY  SEVERITY  REASON  SINCE  MESSAGE                                                                                         
Cluster/capi-quickstart                                     True                     5h35m                                                                                                   
├─ClusterInfrastructure - OpenStackCluster/capi-quickstart                                                                                                                                   
├─ControlPlane - KamajiControlPlane/capo-cp-test                                                                                                                                             
└─Workers                                                                                                                                                                                    
  └─MachineDeployment/capi-quickstart-md-0                  True                     95m                                                                                                     
    └─3 Machines...                                         True                     5h35m  See capi-quickstart-md-0-7d6977868x45z8w-5xjsg, capi-quickstart-md-0-7d6977868x45z8w-87jjg, ...
```

> Please, notice the `KamajiControlPlane/capo-cp-test` resource which is referring to a Kamaji `TenantControlPlane` resource.

```
NAME           VERSION   STATUS   CONTROL-PLANE ENDPOINT   KUBECONFIG                      DATASTORE   AGE
capo-cp-test   v1.24.0   Ready    3.123.28.236:30001       capo-cp-test-admin-kubeconfig   default     5h38m
```

### PR DESCRIPTION
**Pull request by prometherion**
**What this PR does / why we need it**:

Use-case described in #1589.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #1589.

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Review by lentzi90 — COMMENTED** [review]
This looks fine to me but I would like to get some more eyes on it (so keep the `hold`)
/lgtm

**Comment by prometherion** [comment]
Also, I'd like to have your opinion on [this](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1589#issuecomment-1589665315) too, I could update the PR with the required changes to address this.

**Comment by jichenjc** [comment]
/approve


I think the port should also be ok to be changed
if you want separate PR, then please do `hold cancel` to merge this 
or you can combine the effort into same PR, thanks

**Comment by prometherion** [comment]
@mdbooth just for the sake of sharing, the `controlPlaneEndpoint` is not set until the CAPO `OpenStackCluster` has a valid address, backed by a VIP or by a LoadBalancer.

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/bc57b7f750977aebf17ec820cfa5030dc72281f2/controllers/openstackcluster_controller.go#L536-L562

According to my interaction with Kamaji, I was always hitting the error `unable to determine VIP for API server` until the Load Balancer was created externally and reported as `apiServerFixedIP`.

**Comment by mdbooth** [comment]
I had missed that we're only updating the webhook to allow it to be set initially. I'm ok with this.

File: api/v1alpha7/openstackcluster_webhook.go [hunk 1]
```diff
@@ -105,6 +105,11 @@ func (r *OpenStackCluster) ValidateUpdate(oldRaw runtime.Object) error {
 		r.Spec.ControlPlaneEndpoint = clusterv1.APIEndpoint{}
 	}
 
+	// Allow change only for the first time.
+	if old.Spec.DisableAPIServerFloatingIP && old.Spec.APIServerFixedIP == "" {
```
Comment (mdbooth) [inline api/v1alpha7/openstackcluster_webhook.go:109]: Please can you add a unit test for this?

Comment (prometherion) [reply api/v1alpha7/openstackcluster_webhook.go:109]: Implemented with cb025df4.

**Review by mdbooth — COMMENTED** [review]
/lgtm
